src/Microsoft.Azure.EventHubs/EventHubsDiagnosticSource.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.EventHubs
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Diagnostic Source util for "Microsoft.Azure.EventHubs" source.
+    /// </summary>
+    internal static class EventHubsDiagnosticSource
+    {
+        public const string DiagnosticSourceName = "Microsoft.Azure.EventHubs";
+
+        public const string SendActivityName = DiagnosticSourceName + ".Send";
+        public const string SendActivityStartName = SendActivityName + ".Start";
+        public const string SendActivityExceptionName = SendActivityName + ".Exception";
+
+        public const string ReceiveActivityName = DiagnosticSourceName + ".Receive";
+        public const string ReceiveActivityStartName = ReceiveActivityName + ".Start";
+        public const string ReceiveActivityExceptionName = ReceiveActivityName + ".Exception";
+
+        internal static readonly DiagnosticListener DiagnosticListener = new DiagnosticListener(DiagnosticSourceName);
+
+        internal static bool IsEnabled => DiagnosticListener.IsEnabled();
+
+        internal static Activity StartSendActivity(string clientId, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, int count)
+        {
+            // skip if diagnostic source not enabled
+            if (!IsEnabled)
+            {
+                return null;
+            }
+
+            // skip if no listeners for this "Send" activity 
+            if (!DiagnosticListener.IsEnabled(SendActivityName,
+                new
```

Comment:
IsEnabled  must never be called with anonymous object.
The point is to make tracing as efficient as possible. Anonymous objects have to be parsed with reflection.

IsEnabled only accepts 2 parameters now.
You can consider adding another overloads to isEnabled, creating new public type with all properties or using some existing public type to store them.

Reply:
Yes, I went through these considerations but it makes sense to specify 3 here and I'd rather not use ```Tuple```

By creating an overload do you mean contributing to DiagnosticSource? If so then we'd have to wait with this change for the next released version. I think I'll just skip ```partitionKey``` for now